File: app/src/main/java/network/minter/bipwallet/internal/views/list/multirow/SortableRow.java
package network.minter.bipwallet.internal.views.list.multirow;

import android.support.annotation.NonNull;

/**
 * Stars. 2017
 *
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 */
public class SortableRow<V extends MultiRowAdapter.RowViewHolder, T extends MultiRowContract.Row<V>> implements
        MultiRowContract.Row<V> {

    private T row;
    private int position = 0;
    private int previousPosition = -1;

    public SortableRow(final T row, int position) {
        this.row = row;
        this.position = position;
    }

    public void setPosition(int position) {
        this.previousPosition = this.position;
        this.position = position;
    }

    public boolean isInserted() {
        return previousPosition == -1;
    }

    public T getRow() {
        return row;
    }

    @Override
    public int getItemView() {
        return row.getItemView();
    }

    @Override
    public int getRowPosition() {
        return position;
    }

    @Override
    public boolean isVisible() {
        return row.isVisible();
    }

    @Override
    public void onBindViewHolder(@NonNull V viewHolder) {
        row.onBindViewHolder(viewHolder);
    }

    @Override
    public void onUnbindViewHolder(@NonNull V viewHolder) {

    }

    @NonNull
    @Override
    public Class<V> getViewHolderClass() {
        return row.getViewHolderClass();
    }

    public int getPreviousPosition() {
        return previousPosition;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof MultiRowContract.Row) {
            return ((MultiRowContract.Row) obj).getRowPosition() == getRowPosition()
                    && ((MultiRowContract.Row) obj).getItemView() == getItemView()
                    && ((MultiRowContract.Row) obj).isVisible() == isVisible()
                    && ((MultiRowContract.Row) obj).getViewHolderClass() == getViewHolderClass();
        }

        return row.equals(obj);
    }
}

File: app/src/main/java/network/minter/bipwallet/internal/views/list/multirow/SortableRow.kt
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.internal.views.list.multirow

import network.minter.bipwallet.internal.views.list.multirow.MultiRowAdapter.RowViewHolder

/**
 * minter-android-wallet. 2020
 * @author Eduard Maximovich (edward.vstock@gmail.com)
 */
class SortableRow<V : RowViewHolder, T : MultiRowContract.Row<V>>(
        val row: T, position: Int
) : MultiRowContract.Row<V> {

    private var position = 0
    var previousPosition = -1
        private set

    fun setPosition(position: Int) {
        previousPosition = this.position
        this.position = position
    }

    val isInserted: Boolean
        get() = previousPosition == -1

    override fun getItemView(): Int {
        return row.getItemView()
    }

    override fun getRowPosition(): Int {
        return position
    }

    override fun isVisible(): Boolean {
        return row.isVisible()
    }

    override fun onBindViewHolder(viewHolder: V) {
        row.onBindViewHolder(viewHolder)
    }

    override fun onUnbindViewHolder(viewHolder: V) {}
    override fun getViewHolderClass(): Class<V> {
        return row.getViewHolderClass()
    }

    override fun hashCode(): Int {
        return row.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return if (other is MultiRowContract.Row<*>) {
            other.getRowPosition() == getRowPosition() && other.getItemView() == getItemView() && other.isVisible() == isVisible() && other.getViewHolderClass() == getViewHolderClass()
        } else row == other
    }

    init {
        this.position = position
    }
}